docs/remote-snapshotter.md

```diff
@@ -0,0 +1,65 @@
+
+Rather than add another layer of indirection in the form of a network filesystem, we opted to move the whole snapshotter inside the microVM. The lazy loaded filesystem exists only within the guest and can be mounted as a container rootfs from inside the microVM.
+
+With this change, we run 1 remote snapshotter per microVM. Some infrastructure is needed in firecracker-containerd to expose these dynamically spawned snapshotters to containerd as well as proxy snapshotter requests across the VM boundary. This architecture is shown below.
+
+# Architecture
+
+![Firecracker-containerd with remote snapshotter architecture](img/remote-snapshotter-architecture-diagram.png)
+
+## Definitions
+* __Orchestrator__ - A firecracker-containerd client implementation
+* __firecracker-containerd__ - The firecracker-containerd daemon comprised of the vanilla containerd API as well as the Firecracker control API
+* __demux-snapshotter__ - A proxy snapshotter that receives snapshot requests from containerd and proxies them to a remote snapshotter running inside a specific microVM
+* __firecracker-containerd agent__ - The guest side daemon that manages containers inside a microVM. This component is essentially a containerd shim that implements the Task API.
+* __volume-init__ - A process for mounting volumes
+* __MMDS__ - The Firecracker [microVM metadata store](https://github.com/firecracker-microvm/firecracker/blob/v1.1.0/docs/mmds/mmds-user-guide.md). It is an HTTP service inside the microVM that provides host-supplied metadata similar to [IMDS in EC2](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html).
+* __credential-helper-mmds__ - A [Docker credential helper](https://github.com/docker/docker-credential-helpers) that can read credentials from MMDS. Remote snapshotters that support the Docker credential helper pattern can use this tool to retrieve the credentials needed to pull from the remote registry.
+* __socat__ - [A multipurpose relay](http://www.dest-unreach.org/socat/). It is used to expose unix sockets inside the microVM as vsock services that can be reached from the host.
+* __remote snapshotter__ - The snapshotter that is responsible for lazy loading container images
+
+## New components
+
+At a high level, this architecture adds 4 components:
+1) A remote snapshotter inside the Firecracker rootfs that launches at boot 
+1) Socat to expose the remote snapshotter's socket over vsock
+1) A [demux-snapshotter](https://github.com/firecracker-microvm/firecracker-containerd/blob/913ab8bfc13d4735f3cd62dc6ac5f65e80db4400/snapshotter/README.md) which exposes itself a single snapshotter to containerd, but proxies snapshot requests to the in-VM remote snapshotters as they come and go via vsock
```

Comment:
1/ itself as a

2/ nit: I'm in favor of omitting "as they come and go" to make this "to the in-VM remote snapshotters via vsock"